### PR TITLE
Use real-time sockets in federated execution

### PR DESCRIPTION
**Pull request by erlingrj**
Got a bug report from @lhstrh which I was able to reproduce in the following trivial federated program:

```
target C {
    timeout: 100 msec
}
reactor Src {
    timer t (0, 50 msec)
    output out:int
    reaction(t) -> out {=
        lf_set(out, 42);
    =}
}
reactor Sink {
    input in:int 
    reaction(in) {=
        lf_print("Got %d at %li", in->value, lf_time_physical_elapsed());
    =}
}
federated reactor {
    src = new Src()
    sink = new Sink()

    src.out -> sink.in
}
```

I observed a strange ~40msec latency between Src and Sink. I tracked it down be a communication latency between Src and the RTI. After alot of digging I have found out that there is something called Nagle algorithm, enabled by default, which bundles short TCP messages together to avoid too much network traffic. So it was delaying our messages. In this draft I have just added socket options for disabling with `TCP_NODELAY`. It is still a draft because we can discuss which sockets this should apply for, maybe all? After we decide, lets put creation of sockets in `net_util.c` so we dont litter federate.c with these low-level socket option details.

@lhstrh this PR should fix your issue. You should recompile and reinstall the RTI also based on this PR.

### CODE REVIEW

**Comment by erlingrj** [comment]
We should also investigate whether delayed ACKs can be problematic for us also: https://en.wikipedia.org/wiki/Nagle%27s_algorithm#Interaction_with_delayed_ACK

**Comment by erlingrj** [comment]
The `DistributedStopDecentralized.lf` test now fails. This means either 1) I made a mistake here, 2) We shouldnt use TCP_NODELAY for federat-to-federate sockets or 3) Race condition in federated infrastructure related to stop requests which surface with this change

**Comment by edwardalee** [comment]
Using the debugger, I just found a serious bug in federate.c where a local variable "result" was shadowed. I've pushed a fix to this bug to this branch, but this fix should probably be merged with main regardless.  It is interesting that the bug did not show up until @erlingrj changed the sockets to use TCP_NODELAY. In any case, at least on my machine, `DistributedStopDecentralized.lf` does not deadlock anymore.

File: core/federated/RTI/rti_lib.c [hunk 1]
```diff
@@ -53,6 +54,20 @@ int create_server(int32_t specified_port, uint16_t port, socket_type_t socket_ty
     int socket_descriptor = -1;
     if (socket_type == TCP) {
         socket_descriptor = socket(AF_INET, SOCK_STREAM, 0);
+
+        // Disable Nagle algorithm which bundles together small TCP messages to
+        //  reduce network traffic
+        int flag = 1;
+        int result = setsockopt(socket_descriptor,            /* socket affected */
+                                IPPROTO_TCP,     /* set option at TCP level */
+                                TCP_NODELAY,     /* name of option */
```
Comment (lhstrh) [inline core/federated/RTI/rti_lib.c:63]: This code is copied three times in this PR. Should we just put this in a library function named `int disable_delayed_transmission(int sockfd)` or something like that?

**Comment by lhstrh** [comment]
> Using the debugger, I just found a serious bug in federate.c where a local variable "result" was shadowed. I've pushed a fix to this branch.

I think the bug was in Erling's change, not the original code, right?

**Comment by edwardalee** [comment]
Yes, the bug was not in the original code (I initially thought it might have been, but nothing would have worked).

File: core/federated/net_util.c [hunk 2]
```diff
@@ -54,6 +56,34 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 /** Number of nanoseconds to sleep before retrying a socket read. */
 #define SOCKET_READ_RETRY_INTERVAL 1000000
 
+int create_real_time_tcp_socket_errexit() {
+    int sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (sock < 0) {
+        lf_print_error_and_exit("Could not open TCP socket. Err=%d", sock);
+    }
+    // Disable Nagle algorithm which bundles together small TCP messages to
```
Comment (lhstrh) [inline core/federated/net_util.c:64]:
```suggestion
    // Disable Nagle's algorithm which bundles together small TCP messages to
```

**Comment by erlingrj** [comment]
After Christian's fix of the error reporting system, I will now investigate this test failure and fix it so we can merge it soon.

**Comment by erlingrj** [comment]
@lhstrh, I got a test failure on a decentralized federated program on macOS.  No debug output. Unable to reproduce in Ubuntu. Went away when re-running the tests. I have taken a close look at this PR and only  reason I can come up with is either:
1) There is a race condition in decentralized federated which sometimes is triggered now because we have less delays
2) Some unrelated CI issue is at fault.

**Comment by lhstrh** [comment]
My bet would be on there being something fishy with the coordination... I don't think we can blame CI for this one (nor the changes in this PR).

**Comment by erlingrj** [comment]
Then I will merge this and we will have to look at federated race conditions at another time

**Pull request merged**